[DOCS] Update CONTRIBUTING.md, DEVELOPING.md (#1875)
elastic/connectors: docs/CONTRIBUTING.md

3. Declare your dependencies in [requirements.txt](../requirements/framework.txt). Make sure you pin these dependencies.
4. For each dependency you add (including indirect dependencies) list all licences and provide the list in your patch.
5. Your test backend needs to return more than 10k documents as this is the default size limit for Elasticsearch pagination. Having more than 10k documents returned from the test backend will help test the connector more thoroughly. 

## Add new connector

> 👩🏻‍🤝‍👨🏿 _Before investing time in developing a connector, create an issue and reach out to our team for initial feedback on the connector and the libraries it uses!_

If you want to add a new connector source, the following requirements are mandatory for the initial patch:

1. Add a module or a directory in [connectors/sources](../connectors/sources)
2. Implement a class that implements the required methods described in `connectors.source.BaseDataSource`

> 3. Add a unit test in [connectors/sources/tests](../connectors/sources/tests) with **+90% coverage**. Test coverage is run as part of [unit tests](https://github.com/elastic/connectors/blob/main/docs/DEVELOPING.md#testing-the-connector). Look for your file at the end of the console output.

Reviewer: In our `make test` we expect to have more than `92%` test coverage

You can also find us in the `#enterprise-search` channel of the [Elastic Community Slack](http://elasticstack.slack.com).

## Types of contributions

> ### Customize a connector

Reviewer 2: My 2 cents, we will never merge a "contribution" that is a connector customization. If it's a generic improvement, I'd think that falls under "enhancement." If it's specific to the customer's own use case, they're welcome to make changes in their fork, but we will not consider customizations for contributing back. WDYT?

Author: I think we can remove the the category of "Types of contributions" entirely and just have separate sections for "Add new connector" and "Customize a connector". We can incorporate your point about not merging customizations as preamble to the customization option, along with any support-relevant info.

Reviewer 2: > Right, what's the deal with Elastic supporting customized connectors?

I think in that case we support the _framework_ but no longer the connector itself? CC @danajuratoni @ppf2 

> I think we can remove the the category of "Types of contributions" entirely and just have separate sections for "Add new connector" and "Customize a connector". 

I'm fine with that, but at that point it might not make sense for this to be a `CONTRIBUTING.md`. Usually I see these files used _only_ to express to folks what the expectations are if you're going to file issues or PRs. Personally, I'd put stuff for customizing connectors under DEVELOPING.md, or something else. But I think we have a lot of options here, and my opinions aren't super strongly held.

Author: No worries. You know far better than me where to draw the line between contributing guidelines and developing guidelines, we can easily move the customization stuff to DEVELOPING.md

Reviewer: I agree with Sean that this section sounds like: "Adapt existing connector to my specific needs". In that case we shouldn't treat is as contribution and the point about required unit tests % is not necessary. I would also move it away from CONTRIBUTING, Developing sounds ok!

Reviewer 3: I view customized connector clients, converted native connectors (and then customized), and custom connectors in the same category - they are not supported by Elastic.  Users will have to first demonstrate that the issue is occurring in a non-customized connector.   If the user believes that there is a bug in the connector framework, we will need an isolated test case demonstrating that the issue is in the framework, not the customized connector.

To customize an _existing_ connector, follow these steps:

1. Customize the source file for your data source from [connectors/sources](../connectors/sources)

> 2. Add a unit test in [connectors/sources/tests](../connectors/sources/tests) with **+90% coverage**

Reviewer 2: Should we link to a testing reference, or explain how to validate your coverage metrics using `make test`?

Reviewer 2: How about something like:
```
Test coverage is run as part of [unit tests](https://github.com/elastic/connectors/blob/main/docs/DEVELOPING.md#testing-the-connector). Look for your file at the end of the console output.
```
?

3. Declare your dependencies in [requirements.txt](../requirements/framework.txt). Make sure you pin these dependencies.
4. For each dependency you add (including indirect dependencies) list all licences and provide the list in your patch.

> 5. When possible, provide a docker image that runs the backend service, so we can test the connector. If you can't provide a docker image, provide the credentials needed to run against an online service.

Reviewer 2: This would be important for a net-new connector, but I'm not sure its necessary for existing data source types.

6. Your test backend needs to return more than 10k documents as this is the default size limit for Elasticsearch pagination. Having more than 10k documents returned from the test backend will help test the connector more thoroughly. 

### Add new connector

> If you want to add a new connector source, following requirements are mandatory for the initial patch:

Reviewer 2: ```suggestion
If you want to add a new connector source, the following requirements are mandatory for the initial patch:
```

not part of your changes, just noticed on read through.

1. Add a module or a directory in [connectors/sources](../connectors/sources)

> 2. Implement a class that implements **all methods** described in `connectors.source.BaseDataSource`

Reviewer 2: ```suggestion
2. Implement a class that implements the required methods described in `connectors.source.BaseDataSource`
```

This seems out of date.

3. Add a unit test in [connectors/sources/tests](../connectors/sources/tests) with **+90% coverage**

> 4. **Declare your connector** in [config.yml](../config.yml) in the `sources` section

Reviewer 2: ```suggestion
4. **Declare your connector** in [config.py](../connectors/config.py) in the `sources` section of `_default_config()`
```

out of date

5. **Declare your dependencies** in [requirements.txt](../requirements/framework.txt). Make sure you pin these dependencies.
6. For each dependency you add (including indirect dependencies) list all licences and provide the list in your patch.

> 7. Make sure you use an **async lib** for your source. If not possible, make sure you don't block the loop.

Reviewer 2: Worth linking to our DEVELOPING.md section on async?

8. When possible, provide a **docker image** that runs the backend service, so we can test the connector. If you can't provide a docker image, provide the credentials needed to run against an online service.
9. Your **test backend** needs to return more than **10k documents** as this is the default size limit for Elasticsearch pagination. Having more than 10k documents returned from the test backend will help test the connector more thoroughly.

> Before investing time developing a connector, you should create an issue and reach out to our team for initial feedback on the connector and the libraries it uses.

Reviewer 2: I think we should put this at the top of this section, and highlight it.
